module/zfs/vdev_raidz.c

```diff
@@ -1525,41 +1537,70 @@ vdev_raidz_io_start_write(zio_t *zio, raidz_row_t *rr, uint64_t ashift)
 {
 	vdev_t *vd = zio->io_vd;
 	raidz_map_t *rm = zio->io_vsd;
-	int c, i;
+	int c, i = 0, skip_first_cols = -1;
 
 	vdev_raidz_generate_parity_row(rm, rr);
 
-	for (int c = 0; c < rr->rr_cols; c++) {
+	IMPLY(rm->rm_nskip > 0, rm->rm_skipstart < rr->rr_scols);
+
+	if (rr->rr_scols < (rm->rm_skipstart + rm->rm_nskip)) {
+		skip_first_cols =
+		    (rm->rm_skipstart + rm->rm_nskip) % rr->rr_scols;
+	}
+
+	ASSERT3S(skip_first_cols, <, rr->rr_scols);
+	IMPLY(rm->rm_nskip == 0, skip_first_cols = -1);
+
+	for (c = 0; c < rr->rr_scols; c++) {
+		abd_t *abd = NULL;
+		enum zio_flag flags = 0;
 		raidz_col_t *rc = &rr->rr_col[c];
-		if (rc->rc_size == 0)
-			continue;
 
 		/* Verify physical to logical translation */
 		vdev_raidz_io_verify(vd, rr, c);
 
-		zio_nowait(zio_vdev_child_io(zio, NULL,
-		    vd->vdev_child[rc->rc_devidx], rc->rc_offset,
-		    rc->rc_abd, rc->rc_size, zio->io_type, zio->io_priority,
-		    0, vdev_raidz_child_done, rc));
-	}
+		/*
+		 * Generate I/O for skip sectors to improve aggregation
+		 * contiguity. We will use gang ABD's to reduce contention
+		 * on the children VDEV queue locks (vq_lock) by issuing
+		 * a single I/O that contains the data and skip sectors.
+		 */
+		if (((skip_first_cols > -1 && c < skip_first_cols) ||
+		    c >= rm->rm_skipstart) && i < rm->rm_nskip) {
+			abd = abd_alloc_gang();
+			if (rc->rc_size > 0) {
+				abd_gang_add(abd, rc->rc_abd, B_FALSE);
+			} else {
+				ASSERT3P(rc->rc_abd, ==, NULL);
+				flags = ZIO_FLAG_OPTIONAL | ZIO_FLAG_NODATA;
```

Comment:
Why not just skip gang allocation here completely if we are setting ZIO_FLAG_NODATA?

Reply:
We could if we wanted to. Just need to add a little logic to make it happen.

Reply:
Went ahead and update this. It really did not make sense to allocate an ABD in the case of `ZIO_FLAG_NODATA` as you stated. This also removed the need for modifications to `vdev_queue.c`.